fms/testing/fms_hf_equivalence.py

```diff
@@ -0,0 +1,314 @@
+    )
+    return model, tokenizer
+
+
+def load_fms_model(model_path, model_arch, model_variant):
+    tokenizer = AutoTokenizer.from_pretrained(model_path, padding_side="left")
+    model_config_overrides = {"max_expected_seq_len": max_length}
+
+    # TODO hack for now, as llama fms impl sets it to -1, and that causes a crash
+    if model_arch == "llama":
+        model_config_overrides["pad_token_id"] = 0
+
+    logging.info(f"Loading fms Model")
+    model = get_model(
+        architecture=model_arch,
+        variant=model_variant,
+        model_path=model_path,
+        source="hf",
+        device_type=_device,
+        **model_config_overrides,
+    )
+    model = to_hf_api(model=model, **model_config_overrides)
+    logging.info(f"Model loaded")
+    return model, tokenizer
+
+
+def load_prompts(input_json_path: str) -> List[str]:
+    prompts: List[str] = []
+    with open(input_json_path, "r") as input_file:
+        input_data = json.load(input_file)
+        for e in input_data:
+            prompts.append(e["prompt"])
+    return prompts
+
+
+def remove_paddings(v, pad_token_id):
+    mask = v != pad_token_id
+    return v[mask]
+
+
+def tokens_file_path(dir, model_impl_type):
+    return os.path.join(dir, model_impl_type, "tokens")
+
+
+@dataclass
+class DiffSummary:
+    extra_tokens: int = 0
+    less_tokens: int = 0
+    different_tokens: int = 0
+    total_tokens: int = 0
+    total_sequences: int = 0
+    sequences_with_different_tokens: List[int] = field(default_factory=list)
+    sequences_with_extra_tokens: List[int] = field(default_factory=list)
+    sequences_with_less_tokens: List[int] = field(default_factory=list)
+
+    def toJson(self):
+        return {
+            "tokens_stats": {
+                "total_tokens": self.total_tokens,
+                "extra_tokens": f"{(self.extra_tokens * 100/self.total_tokens):.2f}%",
+                "less_tokens": f"{(self.less_tokens * 100/self.total_tokens):.2f}%",
+                "different_tokens": f"{(self.different_tokens * 100/self.total_tokens):.2f}%",
+            },
+            "sequence_stats": {
+                "total_sequences": self.total_sequences,
+                "sequences_with_extra_tokens": f"{(len(self.sequences_with_extra_tokens) * 100 / self.total_sequences):.2f}%",
+                "sequences_with_less_tokens": f"{(len(self.sequences_with_less_tokens) * 100 / self.total_sequences):.2f}%",
+                "sequences_with_different_tokens": f"{(len(self.sequences_with_different_tokens) * 100 / self.total_sequences):.2f}%",
+            },
+            "sequence_info": {
+                "sequences_with_extra_tokens": self.sequences_with_extra_tokens,
+                "sequences_with_less_tokens": self.sequences_with_less_tokens,
+                "sequences_with_different_tokens": self.sequences_with_different_tokens,
+            },
+        }
+
+
+def compute_diff(generated_tokens_dir):
```

Comment:
might want to return a summary object to be used later

Reply:
Done